lunatrace/bsl/backend/src/workers/queue.ts

```diff
@@ -0,0 +1,190 @@
+ * (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * https://github.com/lunasec-io/lunasec/blob/master/licenses/BSL-LunaTrace.txt
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+import { DeleteMessageCommand, Message, ReceiveMessageCommand, ReceiveMessageCommandOutput } from '@aws-sdk/client-sqs';
+
+import { sqsClient } from '../aws/sqs-client';
+import { getQueueHandlerConfig, getWorkerBucketConfig } from '../config';
+import { SqsQueueConfig, WorkerBucketConfig } from '../types/config';
+import { MessageTypeToActivityLookup } from '../types/sqs';
+import { MaybeErrorVoid } from '../types/util';
+import { newError, newResult } from '../utils/errors';
+import { log } from '../utils/log';
+import { getSqsUrlFromName } from '../utils/sqs';
+import { catchError, threwError } from '../utils/try';
+
+import {
+  createMessageTypeToActivityLookup,
+  processLunaTraceSqsMessage,
+} from './queue-processors/process-luna-trace-sqs-message';
+import { processS3SqsMessage } from './queue-processors/process-s3-sqs-message';
+
+async function loadQueueUrlFromName(queueName: string): Promise<string> {
+  const queueUrl = await catchError(getSqsUrlFromName(sqsClient, queueName));
+
+  if (threwError(queueUrl) || queueUrl.error) {
+    log.error('unable to load queue url', {
+      queueName: queueName,
+    });
+    process.exit(-1);
+  }
+
+  const loadedQueueUrl = queueUrl.res;
+
+  log.info('Resolved url for queue', {
+    queueName: queueName,
+    loadedQueueUrl,
+  });
+  return loadedQueueUrl;
+}
+
+export async function startQueueWorker(): Promise<void> {
+  const queueWorker = new QueueWorker();
+  await queueWorker.setup();
+
+  await queueWorker.listenForMessages();
+}
+
+class QueueWorker {
+  private queueUrl?: string;
+  private queueConfig: SqsQueueConfig;
+  private activityLookup: MessageTypeToActivityLookup | null;
+  private bucketConfig: WorkerBucketConfig;
+
+  constructor() {
+    this.queueConfig = getQueueHandlerConfig();
+    this.activityLookup = null;
+    this.bucketConfig = getWorkerBucketConfig();
+  }
+
+  async setup() {
+    this.queueUrl = await loadQueueUrlFromName(this.queueConfig.queueName);
+
+    this.activityLookup = await createMessageTypeToActivityLookup();
+  }
+
+  async deleteMessage(message: Message, queueUrl: string) {
+    const deleteParams = {
+      QueueUrl: queueUrl,
+      ReceiptHandle: message.ReceiptHandle,
+    };
+    try {
+      const data = await sqsClient.send(new DeleteMessageCommand(deleteParams));
+      log.info('Message deleted', data);
+    } catch (err) {
+      log.error('Error deleting message', err);
+    }
+  }
+
+  // processMessageAndReturnResults determines how to process a message from a queue
+  async processMessageAndReturnResults(parsedMessage: any): Promise<MaybeErrorVoid[]> {
+    if ('type' in parsedMessage) {
```

Comment:
Are we trusting amazon to not also use that key and break us?  Maybe its fine but a glance it looks brittle. Should we maybe include a special key like `_isLunasecMessage: true`? Nice comments

Reply:
Interesting point, but if `type` is being used by another amazon data source, then that still works in this design. The person hooking up with data source would add another type to the list of types that are recognized by the message processor and everything should work as expected.